### PR TITLE
Server cli

### PR DESCRIPTION
**Pull request by benjigoldberg**
My first ever rust project. Curious to hear your thoughts.

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -2,9 +2,11 @@ CARGO_HOME=$(HOME)/.cargo
 CARGO_BIN_DIR=$(CARGO_HOME)/bin
 CARGO_WATCH_DIR=$(CARGO_BIN_DIR)/cargo-watch
 CARGO_SYSTEMFD_DIR=$(CARGO_BIN_DIR)/systemfd
+CARGO_CLIPPY_DIR=$(CARGO_BIN_DIR)/cargo-clippy
```
Comment (wilsoniya) [inline Makefile:5]: it's handy to add the cargo bin dir to ur PATH. E.g. my PATH contains: `/home/wilsoniya/.cargo/bin`

File: src/bin/main.rs [hunk 2]
```diff
@@ -0,0 +1,61 @@
+extern crate clap;
```
Comment (wilsoniya) [inline src/bin/main.rs]: FYI it's not necessary to `extern` crates anymore, with the exception being when you want to bring external macros into scope.

File: src/bin/main.rs [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+            server::run(
+                server_matches
+                    .value_of("address")
+                    .expect("should always be populated"),
+                fp_state,
+            )
+            .await?;
+        }
+        _ => {
+            println!("fireplace state set to flame: {}, fan: {}", flame, fan);
```
Comment (wilsoniya) [inline src/bin/main.rs]: you should check out the `log` and `env_logger` crates. For that matter check out adding logging middleware to actix-web (it's a one-liner).

File: src/controllers.rs [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+        };
+        state.fan_dir = Fireplace::_get_direction(state.fan_pin)?;
+        state.flame_dir = Fireplace::_get_direction(state.fan_pin)?;
+        Ok(state)
+    }
+
+    // Given a state, set the fireplace to match the struct's state
+    pub fn set(&mut self, fan: bool, flame: bool) -> Result<(), sysfs_gpio::Error> {
+        self.fan_pin
+            .with_exported(|| Fireplace::_set(self.fan_pin, fan))?;
```
Comment (wilsoniya) [inline src/controllers.rs]: FYI you can do `Self::_set(...)`

File: src/controllers.rs [hunk 5]
```diff
@@ -0,0 +1,60 @@
+extern crate sysfs_gpio;
```
Comment (wilsoniya) [inline src/controllers.rs]: `extern`s should go in the crate root, in your case `lib.rs`. You'll only ever need to `extern` a given crate once for your whole crate.

File: src/controllers.rs [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+
+// GPIO_FAN_PIN is the pin on the raspberry pi that controls the fan
+const GPIO_FAN_PIN: u64 = 23;
+// GPIO_FLAME_PIN is the pin on the raspberry pi that controls the flame
+const GPIO_FLAME_PIN: u64 = 24;
+
+// Fireplace describes the current fireplace condition
```
Comment (wilsoniya) [inline src/controllers.rs]: FYI rustdoc I believe only picks up doc comments prefixed with `///`

File: src/views.rs [hunk 7]
```diff
@@ -0,0 +1,36 @@
+use crate::server;
+use actix_web::error;
+use actix_web::{web, HttpResponse, Responder};
+use serde_derive::Deserialize;
+use std::sync::{Arc, Mutex};
+
+#[derive(Deserialize)]
+pub struct FireplaceFormData {
+    status: String,
```
Comment (wilsoniya) [inline src/views.rs:10]: you could probably make an enum to represent `status`

File: src/views.rs [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+        "fireplaceOn" => (false, true),
+        "allOff" => (false, false),
+        _ => {
+            return Err(error::ErrorBadRequest(format!(
+                "unknown status: {}",
+                form.status
+            )))
+        }
+    };
+    let fireplace_state = data.lock().unwrap();
```
Comment (wilsoniya) [inline src/views.rs]: an option to eliminate unwraps: 

```rust
data.lock()
    .map_err(|_| error::ErrorInternalServerError("Internal Server Error"))
    .and_then(|fireplace_state| {
        fireplace_state.fireplace.lock()
            .map_err(|_| error::ErrorInternalServerError("Internal Server Error"))
            .and_then(|mut fireplace_controller| {
                fireplace_controller.set(fan, flame)
                    .map_err(|e| error::ErrorBadRequest(format!("Unable to set fireplace state: {}", e)))
                    .map(|_| HttpResponse::Ok())
            })
    })

```

Comment (benjigoldberg) [reply src/views.rs]: Nice, I had a lot of fun with this. I actually ended up just wiring the whole function into a map/map_err with the suggested enum change.

**Comment by benjigoldberg** [comment]
@wilsoniya thanks again for taking the time to check this out, you had some really great feedback and I learned a lot 🙇

**Pull request merged**